extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/indexing/MSQCompactionRunner.java

```diff
@@ -415,51 +432,114 @@ private static boolean isQueryGranularityEmptyOrNone(DataSchema dataSchema)
+                                                 .map(DimensionSchema::getName)
+                                                 .collect(Collectors.toSet());
+      if (dataSchema instanceof CombinedDataSchema &&
+          ((CombinedDataSchema) dataSchema).getMultiValuedColumnsInfo().isProcessed()) {
+        // Filter actual MVDs from schema info.
+        Set<String> multiValuedColumnsFromSchema =
+            ((CombinedDataSchema) dataSchema).getMultiValuedColumnsInfo().getMultiValuedColumns();
+        multiValuedColumns = multiValuedColumns.stream()
+                                               .filter(multiValuedColumnsFromSchema::contains)
+                                               .collect(Collectors.toSet());
+      }
+
+      for (String dim : multiValuedColumns) {
+        String virtualColumnExpr = StringUtils.format("mv_to_array(\"%s\")", dim);
+        inputColToVirtualCol.put(
+            dim,
+            new ExpressionVirtualColumn(
+                ARRAY_VIRTUAL_COLUMN_PREFIX + dim,
+                virtualColumnExpr,
+                ColumnType.STRING_ARRAY,
+                injector.getInstance(ExprMacroTable.class)
+            )
+        );
+      }
     }
-    return VirtualColumns.create(new ExpressionVirtualColumn(
-        TIME_VIRTUAL_COLUMN,
-        virtualColumnExpr,
-        ColumnType.LONG,
-        new ExprMacroTable(ImmutableList.of(new TimestampFloorExprMacro(), new TimestampParseExprMacro()))
-    ));
+    return inputColToVirtualCol;
   }
 
-  private static Query<?> buildGroupByQuery(CompactionTask compactionTask, Interval interval, DataSchema dataSchema)
+  private Query<?> buildGroupByQuery(
+      CompactionTask compactionTask,
+      Interval interval,
+      DataSchema dataSchema,
+      Map<String, VirtualColumn> inputColToVirtualCol
+  )
   {
     DimFilter dimFilter = dataSchema.getTransformSpec().getFilter();
 
+    VirtualColumns virtualColumns = VirtualColumns.create(new ArrayList<>(inputColToVirtualCol.values()));
+
+    // Convert MVDs converted to arrays back to MVDs, with the same name as the input column.
+    // This is safe since input column names no longer exist at post-aggregation stage.
+    List<PostAggregator> postAggregators =
+        inputColToVirtualCol.entrySet()
+                            .stream()
+                            .filter(entry -> !entry.getKey().equals(ColumnHolder.TIME_COLUMN_NAME))
+                            .map(
+                                entry ->
+                                    new ExpressionPostAggregator(
+                                        entry.getKey(),
+                                        StringUtils.format("array_to_mv(\"%s\")", entry.getValue().getOutputName()),
+                                        null,
+                                        ColumnType.STRING,
+                                        injector.getInstance(ExprMacroTable.class)
```

Comment:
Rather than using `injector.getInstance()`, inject this field use `@JacksonInject` in the constructor.

Reply:
Done. Thanks!